src/main/java/org/mastodon/mamut/feature/dimensionalityreduction/umap/util/UmapInputDimension.java

```diff
@@ -49,36 +90,90 @@ public String toString()
 	{
-		List< UmapInputDimension< V > > umapInputDimensions = new ArrayList<>();
-		Collection< Feature< V > > features = FeatureUtils.collectFeatureMap( featureModel, vertexType ).values();
-		Collection< Class< ? extends Feature< V > > > excludedFeatures = new ArrayList<>();
-		excludedFeatures.add( Cast.unchecked( SpotTrackIDFeature.class ) );
-		excludedFeatures.add( Cast.unchecked( SpotBranchIDFeature.class ) );
-		for ( Feature< V > feature : features )
+		List< UmapInputDimension< V, E > > umapInputDimensions = new ArrayList<>();
+		Collection< Feature< V > > vertexFeatures = FeatureUtils.collectFeatureMap( featureModel, vertexType ).values();
+		Collection< Class< ? extends Feature< V > > > excludedVertexFeatures = new ArrayList<>();
+		excludedVertexFeatures.add( Cast.unchecked( SpotTrackIDFeature.class ) );
+		excludedVertexFeatures.add( Cast.unchecked( SpotBranchIDFeature.class ) );
+		for ( Feature< V > feature : vertexFeatures )
 		{
-			if ( excludedFeatures.contains( feature.getClass() ) || feature instanceof AbstractUmapFeature )
+			if ( excludedVertexFeatures.contains( feature.getClass() ) || feature instanceof AbstractUmapFeature )
 				continue;
 			for ( FeatureProjection< V > projection : feature.projections() )
-				umapInputDimensions.add( new UmapInputDimension<>( feature, projection ) );
+				umapInputDimensions.add( UmapInputDimension.fromVertexFeature( feature, projection ) );
+		}
+		return umapInputDimensions;
+	}
+
+	private static < V extends Vertex< E >, E extends Edge< V > > List< UmapInputDimension< V, E > >
+			getEdgeDimensions( final FeatureModel featureModel, final Class< E > edgeType )
+	{
+		List< UmapInputDimension< V, E > > umapInputDimensions = new ArrayList<>();
+		Collection< Feature< E > > edgeFeatures = FeatureUtils.collectFeatureMap( featureModel, edgeType ).values();
+		Collection< Class< ? extends Feature< E > > > excludedEdgeFeatures = new ArrayList<>();
+		excludedEdgeFeatures.add( Cast.unchecked( LinkTargetIdFeature.class ) );
+		for ( Feature< E > feature : edgeFeatures )
+		{
+			if ( excludedEdgeFeatures.contains( feature.getClass() ) )
+				continue;
+			for ( FeatureProjection< E > projection : feature.projections() )
+				umapInputDimensions.add( UmapInputDimension.fromEdgeFeature( feature, projection ) );
 		}
 		return umapInputDimensions;
 	}
+
+	private static < V extends Vertex< E >, E extends Edge< V > > ToDoubleFunction< V > edgeProjectionFunction(
+			final FeatureProjection< E > edgeProjection
+	)
+	{
+		return vertex -> {
+			if ( vertex.incomingEdges().isEmpty() )
+				return Double.NaN;
+			else if ( vertex.incomingEdges().size() == 1 )
+				return edgeProjection.value( vertex.incomingEdges().get( 0 ) );
+			double[] edgeValues = new double[ vertex.incomingEdges().size() ];
+			for ( int i = 0; i < edgeValues.length; i++ )
+			{
+				E edge = vertex.incomingEdges().get( i );
+				edgeValues[ i ] = edgeProjection.value( edge );
+			}
+			return Util.average( edgeValues );
```

Comment:
The performance of this code could be improved. It is possible to avoid memory allocations.
The easiest would be to replace `vertex.incomingEdges().get( 0 )` with `vertex.incomingEdges().iterator().next()`.
The other options is to pass a `ref` to `vertex.incomingEdges().get( i )`.

Reply:
Good suggestion. Changed it.